src/cookies.ts

```diff
@@ -0,0 +1,432 @@
+  options: {
+    cookieEncoding: "raw" | "base64url";
+    cookies?:
+      | CookieMethodsBrowser
+      | CookieMethodsBrowserDeprecated
+      | CookieMethodsServer
+      | CookieMethodsServerDeprecated;
+    cookieOptions?: CookieOptionsWithName;
+  },
+  isServerClient: boolean,
+) {
+  const cookies = options.cookies ?? null;
+  const cookieEncoding = options.cookieEncoding;
+
+  const setItems: { [key: string]: string } = {};
+  const removedItems: { [key: string]: boolean } = {};
+
+  let getAll: (keyHints: string[]) => ReturnType<GetAllCookies>;
+  let setAll: SetAllCookies;
+
+  if (cookies) {
+    if ("get" in cookies) {
+      // Just get is not enough, because the client needs to see what cookies
+      // are already set and unset them if necessary. To attempt to fix this
+      // behavior for most use cases, we pass "hints" which is the keys of the
+      // storage items. They are then converted to their corresponding cookie
+      // chunk names and are fetched with get. Only 5 chunks are fetched, which
+      // should be enough for the majority of use cases, but does not solve
+      // those with very large sessions.
+
+      const getWithHints = async (keyHints: string[]) => {
+        // optimistically find the first 5 potential chunks for the specified key
+        const chunkNames = keyHints.flatMap((keyHint) => [
+          keyHint,
+          ...Array.from({ length: 5 }).map((_, i) => `${keyHint}.${i}`),
+        ]);
+
+        const chunks: ReturnType<GetAllCookies> = [];
+
+        for (let i = 0; i < chunkNames.length; i += 1) {
+          const value = await cookies.get(chunkNames[i]);
+
+          if (!value && typeof value !== "string") {
+            continue;
+          }
+
+          chunks.push({ name: chunkNames[i], value });
+        }
+
+        // TODO: detect and log stale chunks error
+
+        return chunks;
+      };
+
+      getAll = async (keyHints: string[]) => await getWithHints(keyHints);
+
+      if ("set" in cookies && "remove" in cookies) {
+        setAll = async (setCookies) => {
+          for (let i = 0; i < setCookies.length; i += 1) {
+            const { name, value, options } = setCookies[i];
+
+            if (value) {
+              await cookies.set!(name, value, options);
+            } else {
+              await cookies.remove!(name, options);
+            }
+          }
+        };
+      } else if (isServerClient) {
+        setAll = async () => {
+          console.warn(
+            "@supabase/ssr: createServerClient was configured without set and remove cookie methods, but the client needs to set cookies. This can lead to issues such as random logouts, early session termination or increased token refresh requests. If in NextJS, check your middleware.ts file, route handlers and server actions for correctness. Consider switching to the getAll and setAll cookie methods instead of get, set and remove which are deprecated and can be difficult to use correctly.",
```

Comment:
If someone sees this in logs, they really need to fix their implementation because they would see random logouts. Maybe I should add an official warning suppression flag?

Reply:
Will add this in a follow up PR.